### PR TITLE
client refactor: simplify mirrors loop

### PR DESCRIPTION
**Pull request by sechkova**
Addresses: #1318, #1339 

**Description of the changes being introduced by the pull request**:

This set of changes tries to simplify the code and make it easy to follow as a result it:

- Removes  mirror_*_download functions and moves the mirror's loop inside the Updater
- Replaces file objects used as function arguments for metadata files, which are read in memory right after download, anyway. Temporary files are now handled in a single function.
- Does small code cleanup

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

### CODE REVIEW

**Comment by jku** [comment]
One of the reasons for using a file object (a streaming object if you will) is that we have not needed to read the file in memory in all cases. With metadata that's a little moot as we do have to read it to verify and process... but with targets _not_ loading the whole file is a very good idea. So I think the the file object commit is not good. This also means the updater_rework _check_hashes() is wrong.

**Comment by jku** [comment]
This is not really a review (sorry) but maybe background for future discussion:

Moving the download loop back to updater looks like the right move (in other words my original idea was bad): I can understand this code :) . However I'm not sure the idea of separate loops is better than what the current/old updater does. The download loop is not trivial (we've had multiple bugs with file objects etc) so is multiplying it going to help?

I think the underlying idea with your design is that 
*  in current updater is very hard to follow how e.g. specific metadata gets verified (what actually gets checked before new root metadata is accepted)
* this becomes easier to understand when the root download code is in its own function and not intermingled with e.g. snapshot download code

This is valid... but I think the cost of duplicating the download code  is too high -- this is visible in root download where you already had to refactor that one to another function. I think having a generic metadata download function is fine we just have to figure out how to make it clear what is verified when: I'm hoping #1313 will help there -- my vision of what would happen is something like this:
 * root download loop starts going through versions N+1
   * it calls generic metadata download loop that start downloading files from mirrors 
     * for each candidate file, it tries to create a new metadata object: this verifies the internal state
     * and tries to insert the object into the current valid metadata bundle (aka local repository controller) as new root: this does most(?) of the required tuf spec checks in the root update part
     * if anything fails we continue the loop,  otherwise our metadata bundle is now using the new root and we can break the loop



Another observation:
* the Updater._verify_*() functions are named confusingly: they create a Metadata object, then they do additional verification and **return the object**. It's not a verifying method, it's a factory method to create a Metadata object that meets some criteria

File: tuf/client_rework/updater_rework.py [hunk 1]
```diff
@@ -213,97 +209,157 @@ def _load_root(self) -> None:
...
+            # than the expiration timestamp in the trusted root metadata file
+            # TODO define which exceptions are part of the public API
+            intermediate_root.expires()
+
+            # 1.9. If the timestamp and / or snapshot keys have been rotated,
+            # then delete the trusted timestamp and snapshot metadata files.
+            if self._metadata["root"].keys(
+                "timestamp"
+            ) != intermediate_root.keys("timestamp"):
+                # FIXME: use abstract storage
```
Comment (MVrachev) [inline tuf/client_rework/updater_rework.py:267]: Could we use TODO instead of FIXME?
I did a quick search in the tuf repo and found only 1 place in the code, besides this file, where we are using those.
It seems we aren't using FIXME in our code and if (hopefully) one day we decide to fix our TODOs this will be left out.

**Comment by MVrachev** [comment]
> This is valid... but I think the cost of duplicating the download code is too high -- this is visible in root download where you already had to refactor that one to another function. I think having a generic metadata download function is fine we just have to figure out how to make it clear what is verified when: I'm hoping #1313 will help there -- my vision of what would happen is something like this:

I was going to write the same thing and then I saw your comment.
I also feel like there is too much repetition and I hope we find another way to do this.
Will follow your `Metadata bundle` suggestion.

**Comment by sechkova** [comment]
Agree that now when mirror download code is moved back to Updater the code is more readable but also makes the repetition obvious. However replacing it with a single function with the current state of the Updater code does not result in the desired clean implementation. I propose keeping this changes as a base to test the work proposed in #1313, see how it fits in and at the same time reduce the code repetition.
I force pushed, fixed a bug that @jku noticed in the mirrors loop and also split the `_check_hashes` function in two: one correctly dealing with file objects for target files download and one using file content read in memory for metadata files.

File: tuf/client_rework/updater_rework.py [hunk 2]
```diff
@@ -147,22 +147,30 @@ def download_target(self, target: Dict, destination_directory: str):
...
-            # TODO: do something with exceptions
-            raise
+                _check_file_length(temp_obj, target["fileinfo"]["length"])
+                temp_obj.seek(0)
+                _check_hashes_obj(temp_obj, target["fileinfo"]["hashes"])
+                break
+
+            except Exception as exception:
+                if temp_obj:
+                    temp_obj.close()
```
Comment (jku) [inline tuf/client_rework/updater_rework.py:174]: this applies to every case with close(): current client does `temp_obj = None` here (inside the if-clause) and it makes sense I think: otherwise temp_obj might be defined on next iteration even if it should not be, leading to us calling close twice... 

Calling close() twice should not break things so this probably does not fail anything but it looks fishy

In fact now that you've made the metadata handling bytes-based (and not file object), all of the metadata download loops could just use a download method that returns bytes -- but it does not need to happen in this PR. Maybe adding temp_obj = None is more straight forward for now?

File: tuf/client_rework/updater_rework.py [hunk 3]
```diff
@@ -147,22 +147,30 @@ def download_target(self, target: Dict, destination_directory: str):
...
+                raise exceptions.NoWorkingMirrorError({file_mirror: exception})
```
Comment (jku) [inline tuf/client_rework/updater_rework.py]: this should store the exception like the other functions, no?

File: tuf/client_rework/updater_rework.py [hunk 4]
```diff
@@ -759,19 +834,34 @@ def _check_file_length(file_object, trusted_file_length):
         )
 
 
-def _check_hashes(file_object, trusted_hashes):
+def _check_hashes_obj(file_object, trusted_hashes):
+    """
+    TODO
+    """
```
Comment (jku) [inline tuf/client_rework/updater_rework.py:855]: just comment: I've filed https://github.com/theupdateframework/tuf/issues/1361 for an easy to use solution for both cases here

File: tuf/client_rework/updater_rework.py [hunk 5]
```diff
@@ -759,19 +834,34 @@ def _check_file_length(file_object, trusted_file_length):
...
+    for algorithm, trusted_hash in trusted_hashes.items():
+        digest_object = sslib_hash.digest_fileobject(file_object, algorithm)
+
+        computed_hash = digest_object.hexdigest()
+
+        # Raise an exception if any of the hashes are incorrect.
+        if trusted_hash != computed_hash:
+            raise sslib_exceptions.BadHashError(trusted_hash, computed_hash)
```
Comment (jku) [inline tuf/client_rework/updater_rework.py]: we have a BadHashError of our own and should not raise sslib errors

**Review by jku — COMMENTED** [review]
So previous comment still stands 

> I think the cost of duplicating the download code is too high

... but we have a plan:
 
* get this code into a readable form again (this PR) even if it keeps duplicating a lot of code
* reduce support for mirrors (as it really is what generates all of this complexity, and no-one needs most of it) : Teodora is going to document that plan in https://github.com/theupdateframework/tuf/issues/1307
* after that we will hopefully be able to simplify this code drastically
* at this point also metadata bundle can be simplified https://github.com/theupdateframework/tuf/pull/1355

After these steps we will be able to evaluate whether the branch has value or if we should just consider it lessons learned and start over...

So with this idea in mind I've reviewed again: I think I found one actual issue in target_download and then some comments...

File: tuf/client_rework/updater_rework.py (hunk 2 above)
Comment (sechkova) [reply tuf/client_rework/updater_rework.py:174]: You're right, thanks for noticing. I added `temp_obj = None` for now, since changing the download function would mean a difference between metadata and targets. I'll keep this in mind.

File: tuf/client_rework/updater_rework.py (hunk 3 above)
Comment (sechkova) [reply tuf/client_rework/updater_rework.py]: Yes, fixed, thanks

**Comment by sechkova** [comment]
Oops my "fixup" commit is not signed off ... but I apparently want to rebase and squash this one after your review. 
On the other hand, do I add the linter fix by running the latest black version in this PR?

**Comment by jku** [comment]
Looks good enough for me -- let's get this merged so we can see if streamlining mirrors helps...

I'll merge develop into experimental-client now (#1365): After that if you'd like to rebase this on experimental-client (to get the linter fix) and also rebase --autosquash (to squash the fixup commit), that works for me.

**Review by jku — APPROVED** [review]

**Comment by sechkova** [comment]
Thanks, it should be ready now.

File: tuf/client_rework/updater_rework.py (hunk 1 above)
Comment (jku) [reply tuf/client_rework/updater_rework.py:267]: I don't really want to use abstract storage in the client (or make it a priority, even if Metadata supports it) but as this part of code is hopefully going to move to bundle anyway, I'm not looking at this sort of things too closely now
